jetstream_pt/engine.py

```diff
@@ -600,23 +602,41 @@ def colocated_cpus(self) -> Union[list[engine_api.CpuDevices], None]:
 
   def get_prefix_destination_sharding(self) -> Prefix:
     """Returns the shardings necessary to transfer data between engines."""
-    return Prefix(
-        self.replicated,
-        self.cache_sharding,
-        self.replicated,
-    )
+    if self.env.shard_on_batch:
+      return Prefix(
+          self.replicated,  # cache is replicated because bs=1 for prefill
+          self.replicated,
+          self.replicated,
+      )
+    else:
+      return Prefix(
+          self.replicated,
+          self.cache_sharding,
+          self.replicated,
+      )
 
   def get_decode_state_sharding(self) -> DecodeState:
     """Gets the shardings corresponding to the decode state."""
-    return DecodeState(
-        self.replicated,
-        self.cache_sharding,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-        self.replicated,
-    )
+    if self.env.shard_on_batch:
+      return DecodeState(
```

Comment:
Nit, can we do 

return DecodeState(
          self.x_sharding if self.env.shard_on_batch else self.replicated,  # shard on batch
          self.cache_sharding,
          self.replicated,
          self.replicated,
          self.replicated,
          self.replicated,
          self.replicated,
      )

Reply:
done.,